datafusion/src/execution/dataframe_impl.rs

```diff
@@ -62,6 +68,65 @@ impl DataFrameImpl {
     }
 }
 
+#[async_trait]
+impl TableProvider for DataFrameImpl {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        Arc::new(Schema::new(
+            self.plan
+                .schema()
+                .fields()
+                .iter()
+                .map(|f| f.field().clone())
+                .collect(),
+        ))
+    }
```

Comment:
I think you can use some rust-fu to make this shorter if you want

```suggestion
    fn schema(&self) -> SchemaRef {
        let schema: Schema = self.plan.schema().as_ref().into();
        Arc::new(schema)
    }

```